src/mlpack/methods/ne/genome.hpp

```diff
@@ -0,0 +1,334 @@
+  bool ContainEnabledLink(int innovId) const {
+    for(int i=0; i < NumLink(); ++i) {
+        if (aLinkGenes[i].InnovationId() == innovId &&
+            aLinkGenes[i].Enabled()) {
+            return true;
+        }
+    }
+    return false;
+  }
+
+  // Whether link exist.
+  bool ContainLink(int innovId) const {
+    for(int i=0; i < NumLink(); ++i) {
+        if (aLinkGenes[i].InnovationId() == innovId) {
+            return true;
+        }
+    }
+    return false;
+  }
+
+  // Set neurons' input and output to zero.
+  void Flush() {
+    for (int i=0; i<aNeuronGenes.size(); ++i) {
+      aNeuronGenes[i].Activation(0);
+      aNeuronGenes[i].Input(0);
+    }
+  }
+
+  // Sort link genes by toNeuron's depth.
+  void SortLinkGenes() {
+    struct DepthAndLink
+    {
+      double depth;
+      LinkGene link;
+ 
+      DepthAndLink(double d, LinkGene& l) : depth(d), link(l) {}
+
+      bool operator < (const DepthAndLink& dL) const
+      {
+        return (depth < dL.depth);
+      }
+    };
+
+    std::vector<double> toNeuronDepths;
+    for (int i=0; i<aLinkGenes.size(); ++i) {
+      NeuronGene toNeuron;
+      GetNeuronById(aLinkGenes[i].ToNeuronId(), toNeuron);
+      toNeuronDepths.push_back(toNeuron.Depth());
+    }
+
+    std::vector<DepthAndLink> depthAndLinks;
+    int linkGenesSize = aLinkGenes.size();
+    for (int i=0; i<linkGenesSize; ++i) {
+      depthAndLinks.push_back(DepthAndLink(toNeuronDepths[i], aLinkGenes[i]));
+    }
+
+    std::sort(depthAndLinks.begin(), depthAndLinks.end());
+
+    for (int i=0; i<linkGenesSize; ++i) {
+      aLinkGenes[i] = depthAndLinks[i].link;
+    }
+  }
+
+  // Activate genome. The last dimension of input is always 1 for bias. 0 means no bias.
+  void Activate(std::vector<double>& input) {
+    assert(input.size() == aNumInput);
+
+    SortLinkGenes();
+    
+    // Set all neurons' input to be 0.
+    for (int i=0; i<NumNeuron(); ++i) {
+      aNeuronGenes[i].Input(0);
+    }
+    
+    // Set input neurons.
+    for (int i=0; i<aNumInput; ++i) {
+      aNeuronGenes[i].Input(input[i]);  // assume INPUT, BIAS, OUTPUT, HIDDEN sequence
+      aNeuronGenes[i].Activation(input[i]);
+    }
+
+    // Activate hidden and output neurons.
+    for (int i = 0; i < NumLink(); ++i) {
+      if (aLinkGenes[i].Enabled()) {
+        int toNeuronIdx = GetNeuronIndex(aLinkGenes[i].ToNeuronId());
+        int fromNeuronIdx = GetNeuronIndex(aLinkGenes[i].FromNeuronId());
+        double input = aNeuronGenes[toNeuronIdx].Input() + 
+                       aNeuronGenes[fromNeuronIdx].Activation() * aLinkGenes[i].Weight();
+        aNeuronGenes[toNeuronIdx].Input(input);
+        
+        if (i == NumLink() - 1) {
+          aNeuronGenes[toNeuronIdx].CalcActivation();
+        } else if (GetNeuronIndex(aLinkGenes[i + 1].ToNeuronId()) != toNeuronIdx) {
+          aNeuronGenes[toNeuronIdx].CalcActivation();
+        }
+      }
+    }
+  }
+
+  // Get output vector.
+  void Output(std::vector<double>& output) {
```

Comment:
I suggest to use arma::vec for the output, in this case, we can use a bunch of mlpack methods that can directly work with the output, what do you think? The same applies to the input. E.g. we could directly use PCA to preprocess the input. Also, we could allocate the memory once, and overwrite the results in every iteration instead of clearing the vector.